---
title: "CNVfilteR: Remove false positives of any CNV calling tool by using SNVs calls "
author: "Jose Marcos Moreno-Cabrera <jmoreno@igtp.cat> and Bernat Gel <bgel@igtp.cat> "
date: "`r doc_date()`"
package: "`r pkg_ver('CNVfilteR')`"
output: 
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{CNVfilteR vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r, include=FALSE}
library(knitr)
opts_chunk$set(concordance=FALSE)
knitr::opts_chunk$set(fig.width = 7)
knitr::opts_chunk$set(fig.height = 7)
set.seed(21666641)
```


# Introduction

Many tools for copy number variant (CNV) detection from NGS data have been 
developed. Usually, these tools were designed for different input data like WGS, WES or
panel data, and their performance may depend on the CNV size. Available
benchmarks show that all these tools obtain false positives, sometimes 
reaching a very high number of them.

With the aim of reducing the number of false positives,
`r BiocStyle::Biocpkg("CNVfilteR")` identifies those CNVs that can be discarded.
This task is performed by using the single nucleotide variant (SNV) calls that are usually 
obtained in common NGS pipelines. As VCF field interpretation is key when working
with these files, `r BiocStyle::Biocpkg("CNVfilteR")` specifically supports 
VCFs produced by VarScan2, Strelka/Strelka2, freeBayes, HaplotypeCaller, and
UnifiedGenotyper.



# Installation

`r BiocStyle::Biocpkg("CNVfilteR")` is a 
[Bioconductor](http://bioconductor.org) package and to install it we have
to use `r BiocStyle::Biocpkg("BiocManager")`.

```{r getPackage, eval=FALSE}
  if (!requireNamespace("BiocManager", quietly = TRUE))
      install.packages("BiocManager")
  BiocManager::install("CNVfilteR")
```


We can also install the package from github to get the latest **devel version**,
but beware that it **might be incompatible with the release version of 
Bioconductor**.

```` {r, eval = FALSE}
  BiocManager::install("jpuntomarcos/CNVfilteR")
```


# Quick Start

Below we show a full example that covers from data loading to results plotting.
First, we load some CNV tool results:
```{r message=FALSE}
library(CNVfilteR)

cnvs.file <- system.file("extdata", "DECoN.CNVcalls.csv", package = "CNVfilteR", mustWork = TRUE)
cnvs.gr <- loadCNVcalls(cnvs.file = cnvs.file, chr.column = "Chromosome", start.column = "Start", end.column = "End", cnv.column = "CNV.type", sample.column = "Sample", genome = "hg19")

```

Then, we load the SNVs stored in a couple of VCF files. 
```{r }
vcf.files <- c(system.file("extdata", "variants.sample1.vcf.gz", package = "CNVfilteR", mustWork = TRUE),
               system.file("extdata", "variants.sample2.vcf.gz", package = "CNVfilteR", mustWork = TRUE))
vcfs <- loadVCFs(vcf.files, cnvs.gr = cnvs.gr)
```
We observe that the function recognized VarScan2 as the source, so fields
were selected and allele frequency consequently. Now we can call 
`filterCNVs()` to identify those CNVs that can be discarded.

```{r message=FALSE }
results <- filterCNVs(cnvs.gr, vcfs)
results
```

And we can check those CNVs that can be filtered out:

```{r}
as.data.frame(results$cnvs[results$cnvs$filter == TRUE])
```

As an example, we can observe that the CNV with id=3 has 4 variants matching 
it: 2 in favor of discarding it, two against discarding it. If we want to know
more about the variants matching a certain CNV we can do:
```{r}
results$variantsForEachCNV[["3"]]
```

Two variants are close to the default expected heterozygous frequency,
0.5, so they obtain a positive score. The other two variants are not 
so clearly close to the default expected duplication value, 0.33, so
they obtain a low negative score. All these default values and others can
be modified in the `filterCNVs()` function.

Finally, we may be interested in plotting the results. For example, we can
plot easily the CNV with id=3 and all the variants matching it.

```{r, fig.width=8, fig.height=4, fig.wide=TRUE}
plotVariantsForCNV(results, "3")
```

On the opposite side, we can observe those CNVs that cannot be discarded:
```{r}
as.data.frame(results$cnvs[results$cnvs$filter != TRUE])
```

For example, the CNV with id=14 has one variant matching it. If we get
the variant info, we see that the variant has an allele frequency very close to
the default expected duplication value, 0.66.

```{r}
results$variantsForEachCNV[["14"]]
```

So, no evidence was found for discarding the CNV. We can also plot the CNV and the variant:
```{r, fig.width=8, fig.height=4, fig.wide=TRUE}
plotVariantsForCNV(results, "14")
```





# Session Info
```{r}
  sessionInfo()
```
